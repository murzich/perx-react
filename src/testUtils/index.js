// eslint-disable-next-line import/prefer-default-export
export { default as renderWithRedux } from './renderWithRedux';
