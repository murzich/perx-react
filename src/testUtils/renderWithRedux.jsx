// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducer from '../state/reducer';

export default function renderWithRedux(
  ui,
  { initialState, store = createStore(reducer, initialState) } = {},
) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    store,
  };
}

export const rerenderWithRedux = rerender => (
  ui,
  { initialState, store = createStore(reducer, initialState) } = {},
) => ({
  ...rerender(<Provider store={store}>{ui}</Provider>),
  store,
});
