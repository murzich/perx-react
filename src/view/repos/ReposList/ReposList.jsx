import PropTypes from 'prop-types';
import React from 'react';
import { List } from 'semantic-ui-react';
import { reposSelectors } from '../../../state/repos';
import { withLoadingState } from '../../hocs';
import { useFirstRendered } from '../../hooks';
import LoadMoreContainer from '../LoadMoreContainer';
import PlaceholderItem from '../PlaceholderItem';
import InitialItem from '../InitialItem';
import ReposItem from '../ReposItem';

const ReposList = ({ repos = [], loading = false }) => {
  const displayedOnce = useFirstRendered(repos, loading);

  return (
    <List divided relaxed>
      {!displayedOnce && !loading && <InitialItem />}

      {loading && (
        <>
          <List.Item>
            <PlaceholderItem />
          </List.Item>
          <List.Item>
            <PlaceholderItem />
          </List.Item>
        </>
      )}

      {!loading && (
        <>
          {repos.map(repo => (
            <ReposItem
              key={repo.id}
              url={repo.html_url}
              name={repo.name}
              description={repo.description}
              stars={repo.stargazers_count}
              watchers={repo.forks_count}
              forks={repo.watchers_count}
              language={repo.language}
            />
          ))}
          <LoadMoreContainer />
        </>
      )}

      {!loading && displayedOnce && repos.length === 0 && (
        <List.Item>
          <List.Content>
            <List.Description>
              {/* eslint-disable-next-line react/no-unescaped-entities */}
              User doesn't have any repository
            </List.Description>
          </List.Content>
        </List.Item>
      )}
    </List>
  );
};

ReposList.propTypes = {
  loading: PropTypes.bool,
  repos: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      html_url: PropTypes.string.isRequired,
      stargazers_count: PropTypes.number.isRequired,
      forks_count: PropTypes.number.isRequired,
      watchers_count: PropTypes.number.isRequired,
      language: PropTypes.string,
    }),
  ).isRequired,
};

export default withLoadingState(reposSelectors.isLoading)(ReposList);
