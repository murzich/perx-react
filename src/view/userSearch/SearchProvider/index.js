import SearchProvider from './SearchProvider';

export default SearchProvider;
