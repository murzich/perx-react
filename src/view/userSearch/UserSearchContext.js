import { createContext } from 'react';

const UserSearchContext = createContext();

export default UserSearchContext;
