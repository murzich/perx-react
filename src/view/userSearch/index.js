import SearchForm from './SearchForm';
import SearchProvider from './SearchProvider';
import UserSearchContext from './UserSearchContext';

export { UserSearchContext, SearchProvider };

export default SearchForm;
