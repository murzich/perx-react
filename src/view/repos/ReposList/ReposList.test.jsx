import { cleanup } from '@testing-library/react';
import React from 'react';
import reposStub from '../../../../__stubs__/repos';
import { renderWithRedux } from '../../../testUtils';
import { rerenderWithRedux } from '../../../testUtils/renderWithRedux';
import { UserSearchContext } from '../../userSearch';
import { InnerComponent as ReposList } from './ReposList';

describe('ReposList', () => {
  afterEach(cleanup);

  const testUser = 'testUser';

  it('should match snapshot when default', () => {
    const { asFragment } = renderWithRedux(
      <UserSearchContext.Provider value={testUser}>
        <ReposList />
      </UserSearchContext.Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot with props', () => {
    const props = {
      repos: reposStub.data,
      loading: false,
    };

    const { asFragment } = renderWithRedux(
      <UserSearchContext.Provider value={testUser}>
        <ReposList {...props} />
      </UserSearchContext.Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot after first update', () => {
    const props = {
      repos: reposStub.data,
      loading: false,
    };

    const { asFragment, rerender } = renderWithRedux(
      <UserSearchContext.Provider value={testUser}>
        <ReposList {...props} />
      </UserSearchContext.Provider>,
    );

    rerenderWithRedux(rerender)(
      <UserSearchContext.Provider value={testUser}>
        <ReposList {...props} repos={[]} />
      </UserSearchContext.Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot with placeholder', () => {
    const props = {
      repos: reposStub.data,
      loading: true,
    };

    const { asFragment } = renderWithRedux(
      <UserSearchContext.Provider value={testUser}>
        <ReposList {...props} />
      </UserSearchContext.Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
