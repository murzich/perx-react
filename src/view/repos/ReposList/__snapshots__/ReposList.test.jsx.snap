// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReposList should match snapshot after first update 1`] = `
<DocumentFragment>
  <div
    class="ui divided relaxed list"
    role="list"
  >
    <div
      class="item"
      role="listitem"
    >
      <div
        class="content"
      >
        <div
          class="description"
        >
          User doesn't have any repository
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ReposList should match snapshot after first update 2`] = `
<DocumentFragment>
  <div
    class="ui divided relaxed list"
    role="list"
  >
    <div
      class="item"
      role="listitem"
    >
      <div
        class="ui placeholder"
      >
        <div
          class="header"
        >
          <div
            class="line short"
          />
        </div>
        <div
          class="paragraph"
        >
          <div
            class="line full"
          />
          <div
            class="line short"
          />
        </div>
        <div
          class="paragraph"
        >
          <div
            class="line medium"
          />
        </div>
      </div>
    </div>
    <div
      class="item"
      role="listitem"
    >
      <div
        class="ui placeholder"
      >
        <div
          class="header"
        >
          <div
            class="line short"
          />
        </div>
        <div
          class="paragraph"
        >
          <div
            class="line full"
          />
          <div
            class="line short"
          />
        </div>
        <div
          class="paragraph"
        >
          <div
            class="line medium"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ReposList should match snapshot when default 1`] = `
<DocumentFragment>
  <div
    class="ui divided relaxed list"
    role="list"
  >
    <div
      class="item"
      role="listitem"
    >
      <div
        class="content"
      >
        <div
          class="header"
        >
          Enter user name to search for his repositories
        </div>
        <div
          class="description"
        >
          There should be a list of all his repositories with public access.
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ReposList should match snapshot with placeholder 1`] = `
<DocumentFragment>
  <div
    class="ui divided relaxed list"
    role="list"
  >
    <div
      class="item"
      role="listitem"
    >
      <div
        class="ui placeholder"
      >
        <div
          class="header"
        >
          <div
            class="line short"
          />
        </div>
        <div
          class="paragraph"
        >
          <div
            class="line full"
          />
          <div
            class="line short"
          />
        </div>
        <div
          class="paragraph"
        >
          <div
            class="line medium"
          />
        </div>
      </div>
    </div>
    <div
      class="item"
      role="listitem"
    >
      <div
        class="ui placeholder"
      >
        <div
          class="header"
        >
          <div
            class="line short"
          />
        </div>
        <div
          class="paragraph"
        >
          <div
            class="line full"
          />
          <div
            class="line short"
          />
        </div>
        <div
          class="paragraph"
        >
          <div
            class="line medium"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ReposList should match snapshot with props 1`] = `
<DocumentFragment>
  <div
    class="ui divided relaxed list"
    role="list"
  >
    <div
      class="item"
      role="listitem"
    >
      <div
        class="content"
      >
        <a
          class="header"
          href="https://github.com/gaearon/6to5"
        >
          6to5
        </a>
        <div
          class="description"
        >
          Turn ES6+ code into readable vanilla ES5 with source maps and more!
        </div>
        <div
          class="ui small labels tags"
        >
          <div
            class="ui label"
          >
            <i
              aria-hidden="true"
              class="star icon"
            />
            1
          </div>
          <div
            class="ui label"
          >
            <i
              aria-hidden="true"
              class="fork icon"
            />
            1
          </div>
          <div
            class="ui label"
          >
            <i
              aria-hidden="true"
              class="eye icon"
            />
            1
          </div>
          <div
            class="ui label"
          >
            JavaScript
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
